docs/src/evaluating.md

```diff
@@ -282,93 +368,130 @@ await page.EvaluateAsync("({ button1, button2 }) => button1.textContent + button
 // Note the required parenthesis.
 await page.EvaluateAsync("([b1, b2]) => b1.textContent + b2.textContent", new[] { button1, button2 });
 
-// Any non-cyclic mix of serializables and handles works.
+// Any mix of serializables and handles works.
 await page.EvaluateAsync("x => x.button1.textContent + x.list[0].textContent + String(x.foo)", new { button1, list = new[] { button2 }, foo = null as object });
 ```
 
-Right:
+## Init scripts
+
+Sometimes it is convenient to evaluate something in the page before it starts loading. For example, you might want to setup some mocks or test data.
+
+In this case, use [`method: Page.addInitScript`] or [`method: BrowserContext.addInitScript`]. In the example below, we will replace `Math.random()` with a constant value.
+
+First, create a `preload.js` file that contains the mock.
+
+```js browser
+// preload.js
+Math.random = () => 42;
+```
+
+Next, add init script to the page.
 
 ```js
-const data = { text: 'some data', value: 1 };
-// Pass |data| as a parameter.
-const result = await page.evaluate(data => {
-  window.myApp.use(data);
-}, data);
+import { test, expect } from '@playwright/test';
+import path from 'path';
+
+test.beforeEach(async ({ page }) => {
+  // Add script for every test in the beforeEach hook.
+  // Make sure to correctly resolve the script path.
+  await page.addInitScript({ path: path.resolve(__dirname, '../mocks/preload.js') });
+});
 ```
 
 ```java
-Map<String, Object> data = new HashMap<>();
-data.put("text", "some data");
-data.put("value", 1);
-// Pass |data| as a parameter.
-Object result = page.evaluate("data => {\n" +
-  "  window.myApp.use(data);\n" +
-  "}", data);
+// In your test, assuming the "preload.js" file is in the "mocks" directory.
+page.addInitScript(Paths.get("mocks/preload.js"));
 ```
 
 ```python async
-data = { 'text': 'some data', 'value': 1 }
-# Pass |data| as a parameter.
-result = await page.evaluate("""data => {
-  window.myApp.use(data)
-}""", data)
+# In your test, assuming the "preload.js" file is in the "mocks" directory.
+await page.add_init_script(path="mocks/preload.js")
 ```
 
 ```python sync
-data = { 'text': 'some data', 'value': 1 }
-# Pass |data| as a parameter.
-result = page.evaluate("""data => {
-  window.myApp.use(data)
-}""", data)
+# In your test, assuming the "preload.js" file is in the "mocks" directory.
+page.add_init_script(path="mocks/preload.js")
 ```
 
 ```csharp
-var data = new { text = "some data", value = 1};
-// Pass data as a parameter
-var result = await page.EvaluateAsync("data => { window.myApp.use(data); }", data);
+// In your test, assuming the "preload.js" file is in the "mocks" directory.
+await Page.AddInitScriptAsync(scriptPath: "mocks/preload.js");
 ```
 
-Wrong:
+######
+* langs: js
```

Comment:
just a thought(feel free to resolve): in theory we could recommend running esbuild as a sub-process on the language ports.